packages/@aws-cdk/aws-ec2/lib/launch-template.ts

```diff
@@ -0,0 +1,688 @@
+
+  /**
+   * @inheritdoc
+   */
+  public readonly launchTemplateId?: string;
+
+  /**
+   * @inheritdoc
+   */
+  public readonly launchTemplateName?: string;
+
+  // =============================================
+  //   Data members
+
+  /**
+   * The default version for the launch template.
+   *
+   * @attribute
+   */
+  public readonly defaultVersionNumber: string;
+
+  /**
+   * The latest version of the launch template.
+   *
+   * @attribute
+   */
+  public readonly latestVersionNumber: string;
+
+  /**
+   * IAM Role assumed by instances that are launched from this template.
+   */
+  public readonly role: iam.IRole;
+
+  /**
+   * Principal to grant permissions to.
+   */
+  public readonly grantPrincipal: iam.IPrincipal;
+
+  /**
+   * Allows specifying security group connections for the instance.
+   *
+   * Note: If the LaunchTemplate was not created with a SecurityGroup, then this connections
+   * object will be empty and operations on it will do nothing.
+   */
+  public readonly connections: Connections;
+
+  /**
+   * UserData executed by instances that are launched from this template.
+   */
+  public readonly userData?: UserData;
+
+
+  protected tagSpecifications: LaunchTemplateTagSpecification[];
+
+  // =============================================
+
+  constructor(scope: Construct, id: string, props?: LaunchTemplateProps) {
+    super(scope, id);
+
+    // Basic validation of the provided spot block duration
+    const spotDuration = props?.spotOptions?.blockDuration?.toHours({ integral: true });
+    if (spotDuration !== undefined && (spotDuration < 1 || spotDuration > 6)) {
+      // See: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-requests.html#fixed-duration-spot-instances
+      Annotations.of(this).addError('Spot block duration must be exactly 1, 2, 3, 4, 5, or 6 hours.');
+    }
+
+    this.role = props?.role ?? new iam.Role(this, 'Role', {
+      assumedBy: new iam.ServicePrincipal('ec2.amazonaws.com'),
+    });
+    this.grantPrincipal = this.role;
+    const iamProfile = new iam.CfnInstanceProfile(this, 'Profile', {
+      roles: [this.role.roleName],
+    });
+
+    if (props?.securityGroup) {
+      this.connections = new Connections({ securityGroups: [props.securityGroup] });
+    } else {
+      // Note: LaunchTemplates are not associated with a VPC, so we cannot create a
+      // SecurityGroup by default. So, we preserve the IConnectable interface by
+      // creating a Connections object but using it when empty will just result in no-ops.
+      this.connections = new Connections({ securityGroups: [] });
+    }
+    const securityGroupsToken = Lazy.list({
+      produce: () => {
+        if (this.connections.securityGroups.length > 0) {
+          return this.connections.securityGroups.map(sg => sg.securityGroupId);
+        }
+        return undefined;
+      },
+    });
+
+    this.userData = props?.userData;
+    if (!this.userData !== undefined && props?.machineImage !== undefined) {
```

Comment:
```ts
!this.userData !== undefined 
```

I'm not sure the precedences work out correctly here.

But isn't the following just as easy:

```ts
this.userData = props.userData ?? props.machineImage?.getImage(this).userData
```